### ISSUE TITLE
Liquid error: undefined method `to_liquid`

### ISSUE THREAD

**Issue by passalini** [Shopify/liquid#470]
I'm with Rails 4 and ruby 2, and when I try to use Liquid with a collection of my objects, Liquid return me `Liquid error: undefined method "to_liquid"`, and to do it works I have to create a method in the object's class:

``` ruby
def to_liquid
  attributes
end
```

I don't know if I'm doing something wrong, but this was the only way that I could access the attributes.

ps: I think that this method should be a default in Liquid gem

**Comment by dbaq** [comment]
Hey guys,

I am running into this issue too. I understood, that it is better to do not expose all the attributes to liquid. But what's the best way to avoid this error when we are dealing with a third party gem? Do I have to wrap these objects and define a to_liquid method just to expose the attributes I want?

``` 
rendered = Liquid::Template.parse("Hi {{shop.name}}").render('shop' => ShopifyAPI::Shop.current)
==> hi Liquid error: undefined method `to_liquid' for #
```

thanks for your help

**Comment by fw42** [comment]
Don't expose your objects to Liquid directly, encapsulate them in Liquid::Drop subclasses and whitelist only the attributes that you intend to be used within the template.

**Comment by misner** [comment]
How exactly should I "not expose your objects to Liquid directly, encapsulate them in Liquid::Drop subclasses and whitelist only the attributes that you intend to be used within the template."

As it seems a very common security issue and best practice ("Exposing all your attributes via Liquid is usually not what you want. You want to whitelist things, not expose everything."), could you add something about practically how to do it on the docucmentation ?

**Comment by dylanahsmith** [comment]
https://github.com/Shopify/liquid/wiki/Introduction-to-Drops

**Comment by misner** [comment]
Thanks Dylan,

Thanks yes I saw actually this page before, but I’m pretty new to RoR and I couldn't practically see how to use it .

What I need is pretty basic :  I have only one column/attribute that I want Liquid to parse and I don't want other atttributes and other tables to be "exposed" (following the advice on this page).

From the comment above saying it’s safer not to expose everything to Liquid, I induced that today, by default, everything is exposed now in my database to Liquid.

But how to restrict it? The sources I found seem contradictory or maybe I am missing sth

-  I saw this first: http://ruby-journal.com/liquid-template-in-ruby-done-right/
Here people use liquid_methods. it seemed pretty simple so I tried (even though I did not see this method in the Wikis of Liquid).

`liquid_method: :name `

I tried it but it did not work, the other columns not explicited on liquid_method seem also still be parsable by Liquid.


- then I read this: https://rubyplus.com/articles/3761-Liquid-Template-in-Rails-5
Here they use something else: to_liquid then use a subclass of Drops.
I tried it also but did not work.
What I don’t get is if, I use a subclass, does that mean that all other models and attrributes that are not mentioned in this subclass, will become "unparsable" by Liquid and hence « secured/safe »/not exposed »?

Sorry if the question seems very basic. Maybe I could write a stackoverflow question, as others might have the same question.

Best
M.

**Comment by pushrax** [comment]
The normal way to do this is like so:

```ruby
class MyModelDrop < Liquid::Drop
  def initialize(my_model)
    @my_model = my_model
  end

  def my_attribute
    @my_model.my_attribute # example
  end

  def another_attribute
    @my_model.settings['another_attribute'].to_s # example
  end
end

class MyModel < ActiveRecord::Base
  def to_liquid
    @drop ||= MyModelDrop.new(self)
  end
end
```

Then you can pass your `MyModel` instance into the liquid arguments for use in templates.

**Comment by misner** [comment]
Thanks very much, I'll try now. 

Just to confirm i understand well: in your example it means all the other attributes of My_model (example my_attribute2, my_attribute3), as they're not present in your code, will NOT be exposed to Liquid, as advised above ("not expose your objects to Liquid directly, encapsulate them in Liquid::Drop subclasses and whitelist only the attributes that you intend to be used within the template.").

You "whitelisted" only 2 attributes: my_attribute and another_attribute.

What about other models apart from my_model ? Are they exposed? Should I also apply your code for all my active record models (even if it means whitelisting inside the code no attribute at all for these other models) if I don't want them exposed ?

**Comment by pushrax** [comment]
By default Liquid doesn't expose anything outside of drops and certain primitives. So, unless you explicitly make things available in drops, they will not be accessible in Liquid. Additionally, unless you pass an instance of a model to Liquid, then Liquid can't access anything from that model. It only operates on things you pass to the `render` function.

So in your case if you really only need one attribute, you could even consider just passing that attribute instead of your whole model instance.

**Comment by misner** [comment]
I see, I think I got it wrong. that's why it was working even though I did not use Drops or "make explicitly make things available in drops".

I think I don't need to whitelist anything as I don't pass instances. As you say I pass only one attribute:
`
@template = Liquid::Template.parse( Modal.where(id: modal['modal_id']).first.modal_content )`

Modal.where(id: modal['modal_id']).first.modal_content looks like this for example:
"This is the {{text1}}<br/>and this is {{image1}} and {{text2}}"

then: I render in the html view

```
<%= sanitize( @template.render('text1' => Deal.content['text1'],
                               'text2' => Deal.content['text2'],                                          
                               'image1' =>Deal.content['image1'])
                                  ) %>
```

(Deal's content attribute is a hash)

So basically I put inside the render only one attribute of Deal Model called 'content'. and then I parse one attribute of Modal model called modal_content.

**I am never passing a "whole" instance** so I am safe I guess.

Thanks for your support. Hopefully will be useful also for other people.
